calcom/cal.com
# feat: cal ai

This PR is a draft, we plan on moving it out of draft by EOD, but are still fixing up issues with app installation.
There also seems to be an issue with the core getEventTypes API in main, where it works in prod but not locally.

## What does this PR do?



Demo: https://www.loom.com/share/950b0f03e1534cad92295e4720440ab4?sid=0eb02880-ea97-4d6a-9586-9582d8eb9caf



This pr adds the Cal AI app to the cal apps folder and the App Store.

The Cal AI app is an email chat interface that allows you to:
- Chat with your calendar: "What meetings do I have today?"
- Book Meetings: "Book a meeting with bob@example.com on Friday at 2pm"
- Update Meetings: "Push the meeting with bob an hour back"
- Delete Meetings: "Cancel the meeting with Bob"
- More complicated commands (subject to vercel function timeout): "Cancel all my meetings this week, push anything with Bob to next week and book a meeting with bob called 'Take me to the hospital' right now"

The Cal AI app is a new application in the /apps folder that should be deployed to https://cal.ai

## Requirement/Documentation



- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change




## How should this be tested?



- ~~Are there environment variables that should be set?~~
- ~~What are the minimal test data to have?~~
- ~~What is expected (happy path) to have (input and output)?~~
- ~~Any other important info that could help to test that PR~~

You will need:

- A running instance of the cal.com monorepo with a db
- An admin access token
- An Open AI account with access to GPT 4
- A Resend Account
- A Sendgrid Account with inbound parse configured

- Populate the .env file in apps/ai based on the .env.example file 

- Run the web, api and ai apps `yarn dev:ai`

- Navigate to the web app and install the cal ai app

- Email the email you configured with inbound parse any instructions

## Performance

Email has inherent speed limitations based on your client.
Apple writes in real time but reads every minute
Gmail reads in real time but writes 10 seconds later because of mandatory “undo send” feature
Additionally, Resend takes some time to send

The AI agent loop can be benchmarked with [langsmith](https://smith.langchain.com)
For example, check out [this run](https://smith.langchain.com/public/67d18947-48c7-47a0-ac95-0f2c45078e03/r)

<img width="312" alt="Screenshot 2023-08-26 at 1 44 24 AM" src="https://github.com/calcom/cal.com/assets/36115192/cabbb185-58ee-4fe0-98b2-4620ef8b380d">

On complex tasks with many steps, vercel can timeout the agent if it goes longer than 1 minute on a pro plan.

## Mandatory Tasks


## Checklist



- ~~I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)~~
- ~~My code doesn't follow the style guidelines of this project~~
- ~~I haven't commented my code, particularly in hard-to-understand areas~~
- ~~I haven't checked if my PR needs changes to the documentation~~
- ~~I haven't checked if my changes generate no new warnings~~
- ~~I haven't added tests that prove my fix is effective or that my feature works~~
- ~~I haven't checked if new and existing unit tests pass locally with my changes~~
